plonky2/src/plonk/conditional_recursive_verifier.rs

```diff
@@ -0,0 +1,352 @@
+use itertools::Itertools;
+use plonky2_field::extension::Extendable;
+
+use crate::fri::proof::{
+    FriInitialTreeProofTarget, FriProofTarget, FriQueryRoundTarget, FriQueryStepTarget,
+};
+use crate::gadgets::polynomial::PolynomialCoeffsExtTarget;
+use crate::hash::hash_types::{HashOutTarget, MerkleCapTarget, RichField};
+use crate::hash::merkle_proofs::MerkleProofTarget;
+use crate::iop::ext_target::ExtensionTarget;
+use crate::iop::target::{BoolTarget, Target};
+use crate::plonk::circuit_builder::CircuitBuilder;
+use crate::plonk::circuit_data::{CommonCircuitData, VerifierCircuitTarget};
+use crate::plonk::config::{AlgebraicHasher, GenericConfig};
+use crate::plonk::proof::{OpeningSetTarget, ProofTarget, ProofWithPublicInputsTarget};
+use crate::with_context;
+
+impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
+    /// Verify `proof0` if `condition` else verify `proof1`.
+    /// `proof0` and `proof1` are assumed to use the same `CommonCircuitData`.
+    pub fn conditionally_verify_proof<C: GenericConfig<D, F = F>>(
+        &mut self,
+        condition: BoolTarget,
+        proof_with_pis0: ProofWithPublicInputsTarget<D>,
+        inner_verifier_data0: &VerifierCircuitTarget,
+        proof_with_pis1: ProofWithPublicInputsTarget<D>,
+        inner_verifier_data1: &VerifierCircuitTarget,
+        inner_common_data: &CommonCircuitData<F, C, D>,
+    ) where
+        C::Hasher: AlgebraicHasher<F>,
+    {
+        let ProofWithPublicInputsTarget {
+            proof:
+                ProofTarget {
+                    wires_cap,
+                    plonk_zs_partial_products_cap,
+                    quotient_polys_cap,
+                    openings,
+                    opening_proof,
+                },
+            public_inputs,
+        } = proof_with_pis0;
+        let ProofWithPublicInputsTarget {
+            proof:
+                ProofTarget {
+                    wires_cap: dummy_wires_cap,
```

Comment:
nit: Let's not call it a dummy proof since technically the method is more general. Seems like it could theoreticaly be used to verify either of two nontrivial circuits, in some particular cases where the two circuits happen to have the same size, gates, etc.